HIP: Add RDNA3 WMMA support to MMF

unverbraucht opened the pull request:
Based on the work by @zhang-hui-yulo for RDNA4 I attempted to backport the WMMA MMF support to RDNA3. Also ports the RDNA4 WMMA-MMQ improvements by @jiachengjason from PR #17156  to RDNA3.

The differences to RDNA4 are:
- RDNA3 has no FP8 support for in WMMA (INT8 is supported by hardware)
- RDNA3 has a different tile size 

The results for granite 1b 400m look great:
| GPU        | Model             |   Microbatch size | Test   |   t/s master |   t/s ba256617d |   Speedup |
|:-----------|:------------------|------------------:|:-------|-------------:|----------------:|----------:|
| RX 7900 XT | granitemoe ?B F16 |                 1 | pp512  |       283.42 |          286.08 |      1.01 |
| RX 7900 XT | granitemoe ?B F16 |                 2 | pp512  |       124.99 |          668.81 |      5.35 |
| RX 7900 XT | granitemoe ?B F16 |                 4 | pp512  |       205.77 |         1224.45 |      5.95 |
| RX 7900 XT | granitemoe ?B F16 |                 8 | pp512  |       377.29 |         1881.51 |      4.99 |
| RX 7900 XT | granitemoe ?B F16 |                16 | pp512  |       640.67 |         3181.89 |      4.97 |
| RX 7900 XT | granitemoe ?B F16 |                32 | pp512  |      1024.92 |         5654.28 |      5.52 |
| RX 7900 XT | granitemoe ?B F16 |                64 | pp512  |      2052.33 |         9817.10 |      4.78 |
| RX 7900 XT | granitemoe ?B F16 |               128 | pp512  |      3622.50 |        15972.81 |      4.41 |
| RX 7900 XT | granitemoe ?B F16 |               256 | pp512  |      6007.40 |        22525.58 |      3.75 |
| RX 7900 XT | granitemoe ?B F16 |               512 | pp512  |      9174.28 |        27815.62 |      3.03 |

EDIT: the performance regression for GPT OSS 20b has been fixed, now we have moderate speed-up:
| GPU        | Model            |   Microbatch size | Test   |   t/s 55ab25caf |   t/s 8aed111ee |   Speedup |
|:-----------|:-----------------|------------------:|:-------|----------------:|----------------:|----------:|
| RX 7900 XT | gpt-oss 20B Q8_0 |                 1 | pp512  |          184.01 |          181.51 |      0.99 |
| RX 7900 XT | gpt-oss 20B Q8_0 |                 2 | pp512  |          194.39 |          216.68 |      1.11 |
| RX 7900 XT | gpt-oss 20B Q8_0 |                 4 | pp512  |          331.38 |          386.97 |      1.17 |
| RX 7900 XT | gpt-oss 20B Q8_0 |                 8 | pp512  |          535.49 |          656.94 |      1.23 |
| RX 7900 XT | gpt-oss 20B Q8_0 |                16 | pp512  |          683.39 |          772.00 |      1.13 |
| RX 7900 XT | gpt-oss 20B Q8_0 |                32 | pp512  |          898.96 |         1049.12 |      1.17 |
| RX 7900 XT | gpt-oss 20B Q8_0 |                64 | pp512  |         1089.26 |         1358.59 |      1.25 |
| RX 7900 XT | gpt-oss 20B Q8_0 |               128 | pp512  |         1712.74 |         1935.43 |      1.13 |
| RX 7900 XT | gpt-oss 20B Q8_0 |               256 | pp512  |         2552.29 |         2828.21 |      1.11 |
| RX 7900 XT | gpt-oss 20B Q8_0 |               512 | pp512  |         3298.97 |         3594.53 |      1.09 |

CC @jiachengjason

Commit:
mmf for rdna4

Commit:
align the padding for rdna4

zhang-hui-yulo committed:
Merge branch 'ggml-org:master' into mmf_wmma_rdna4

Commit:
forbit mul_mat_f for rdna4

zhang-hui-yulo committed:
Merge branch 'ggml-org:master' into mmf_wmma_rdna4

Commit:
fix as comment

Commit:
remove device kernels

Commit:
add constexpr for early return

zhang-hui-yulo committed:
Merge branch 'ggml-org:master' into mmf_wmma_rdna4

Commit:
update based on review comment

Commit:
change based on the review comment

zhang-hui-yulo committed:
Merge branch 'ggml-org:master' into mmf_wmma_rdna4

Commit:
pass compile error

zhang-hui-yulo committed:
Merge branch 'ggml-org:master' into mmf_wmma_rdna4

Commit:
keep code consistency

zhang-hui-yulo committed:
Merge branch 'ggml-org:master' into mmf_wmma_rdna4

zhang-hui-yulo committed:
Merge branch 'ggml-org:master' into mmf_wmma_rdna4

Commit:
Attempt to port RDNA4 WMMA optimizations to RDNA3

  Key Changes Made:

  1. ggml/src/ggml-cuda/common.cuh:
    - Extended AMD_WMMA_AVAILABLE macro to include both RDNA3 and RDNA4
    - Updated amd_wmma_available() to return true for both architectures
  2. ggml/src/ggml-cuda/mma.cuh:
    - Tile structures: Added RDNA3-specific tile sizes:
        - RDNA4: 4 half2 = 8 FP16 elements (compact layout)
      - RDNA3: 8 half2 = 16 FP16 elements (duplicate layout required by hardware)
    - MMA operations: Added RDNA3 intrinsics:
        - FP16: __builtin_amdgcn_wmma_f32_16x16x16_f16_w32 (no gfx12 suffix)
      - BF16: __builtin_amdgcn_wmma_f32_16x16x16_bf16_w32
      - Uses halfx16_t/bf16x16_t for RDNA3 vs halfx8_t/bf16x8_t for RDNA4
    - Load operations: Added conditional handling for 32-byte RDNA3 tiles using two 16-byte copies
  3. ggml/src/ggml-cuda/mmf.cu:
    - Updated to use amd_wmma_available() for both RDNA3 and RDNA4

Commit:
Merge branch 'master' into mmf_wmma_rdna3

Commit:
WMMA RDNA3 fixes

Commit:
fix RDNA3 not using the fast DP4A-based MMQ path. RDNA4 should still use MMQ with integer WMMA operations (hardware-accelerated)

Commit:
more fixes for RDNA3 crashes with quantized models

JohannesGaessler reviewed (commented):
* According to the AMD ISA documentation RDNA3 supports integer tensor cores. Please change the comments to say that it's not implemented rather than not supported.
* Please always add a comment for an `#endif` to indicate which `#if`/`#ifdef` it is closing.
* The `get_i` and `get_j` methods are not going to work correctly if you mirror the data for RDNA3. Please either implement them correctly or replace them with `NO_DEVICE_CODE` for RDNA3.
* The code in `mma.cuh` is currently in a bad state in terms of maintainability and is in dire need of a refactor. However, I consider this to be a job for me, the maintainer, rather than contributors. So no action from your side is necessary, for now it's fine to pile on hacky solutions. I just want to give you a heads-up that the code is subject to change once RDNA3, RDNA4, and CDNA are properly supported and I know what the requirements are.

unverbraucht commented:
@JohannesGaessler thanks for the feedback.

RDNA3 indeed supports INT8 in WMMA, and I'll investigate that. It doesn't support FP8 and the sparse WMMA is also missing. Looking into get_i and get_j.

Regarding your new code in #17505 - does it even make sense to investigate this code here more or should I wait for that PR to be merged and then attempt to add this to the new MMA kernel?

Commit:
Fix RDNA3 WMMA int8 codepath

Details
1. Separated RDNA3 and RDNA4 integer WMMA implementations in mma.cuh:
  - RDNA4: Uses __builtin_amdgcn_wmma_i32_16x16x16_iu8_w32_gfx12 with int32x2_t (original path preserved)
  - RDNA3: Uses __builtin_amdgcn_wmma_i32_16x16x16_iu8_w32 with int32x4_t (new path added)

2. Both architectures now share:                                                                                                                                                       - AMD_WMMA_INT_AVAILABLE macro (enables shared optimizations)
  - Memory layout settings (mmq_x_max=128, granularity=16, nwarps=8)
  - The tile<16, 4, int> optimizations in mmq.cuh
3. RDNA4-exclusive features remain untouched:
  - FP8/BF8 WMMA operations
  - Specific RDNA4 optimizations behind #if defined(RDNA4) guards

Commit:
fix endif commets

JohannesGaessler commented:
>My suggestion will be tile<I, J, T, transposed = true> for matrix A, B for RDNA int8 and tile<I, J, T> for matrix C

Please only do that if the data is actually transposed. As I said, the current state of `mma.cuh` is very messy. I intend to later do a refactor with a better design but I first need to figure out what the exact requirements are. As of right now, if you just need to specify a different data layout for C/D and A/B, preferably do this via the `data_split` enum (may make sense to rename this to `data_layout`).

zhang-hui-yulo commented:
> > My suggestion will be tile<I, J, T, transposed = true> for matrix A, B for RDNA int8 and tile<I, J, T> for matrix C
> 
> Please only do that if the data is actually transposed. As I said, the current state of `mma.cuh` is very messy. I intend to later do a refactor with a better design but I first need to figure out what the exact requirements are. As of right now, if you just need to specify a different data layout for C/D and A/B, preferably do this via the `data_split` enum (may make sense to rename this to `data_layout`).

I prefer data_layout enum including row-major and col-major, all tile for NV shall be row-major. For AMD, matrix A and B shall be row-major (follow NV's pattern) and matrix C shall be col-major (transpose I and J), this makes more sense and more friendly to load_generic.

JohannesGaessler commented:
Agreed, the reason it currently looks like this is that I previously had a different design for Volta. But the performance turned out to be bad so I cut it again.

Commit:
More fixes on get_i/get_j, add support for more tile sizes. get_j respects tile duplication

  1. ✅ Added support for tile<16, 4> and tile<16, 8> in supported()
  2. ✅ Implemented get_i/get_j for these tile sizes
  3. ✅ Set ne = I*J/16 for RDNA3 input tiles (to match 4 VGPR requirement)
  4. ✅ Set ne = I*J/32 for RDNA3 accumulator tiles (same as RDNA4)
  5. ✅ Configured get_j to handle duplication (threads 16-31 load same as 0-15)
  6. ✅ Updated half2/bfloat162 get_j to properly iterate

Commit:
Disable WMMA for RDNA3, enable only for RDNA4

   I could not get the WMMA intrinsics for RDNA3
   (__builtin_amdgcn_wmma_i32_16x16x16_iu8_w32) to work reliably
   wrt egister layout and data handling.

   This change restricts AMD_WMMA_AVAILABLE to RDNA4 only, allowing
   RDNA3 to use the proven fallback MMQ/MFMA code path.

   Removes AMD_WMMA_INT_AVAILABLE macro as it's no longer needed.

JohannesGaessler commented on ggml/src/ggml-cuda/mmq.cuh:
```diff
@@ -256,7 +256,7 @@ static constexpr __device__ int mmq_get_granularity_device(const int /*mmq_x*/)
 #if defined(GGML_USE_HIP)
 static int mmq_get_nwarps_host(const int cc, const int warp_size) {
-    return amd_mfma_available(cc) ? 8 : 256/warp_size;
+    return (amd_mfma_available(cc) || amd_wmma_available(cc)) ? 8 : 256/warp_size;
```
Why are you changing this?

JohannesGaessler reviewed (commented):
Having to deal with `AMD_WMMA_AVAILABLE` vs. `AMD_WMMA_INT8_AVAILABLE` would incur a non-negligible maintenance burden. If it comes to that I would rather not merge this PR and wait for me or someone else to make a PR that properly supports the whole feature set. If I remember correctly you previously had a version that worked correctly and changed only the FP16 and BF16 code paths without touching the int8 code path. A version like that would be acceptable to merge.

zhang-hui-yulo commented:
After https://github.com/ggml-org/llama.cpp/pull/17576, the thing I would do are.
- Add RDNA3 support in common.cuh and mmf.cu like amd_wmma_available, just enable the feature on CPU side.
- Add tile<16,16,half2> and tile<16,16,bfloat162> support for RDNA3 in mma.cuh, the get_i and get_j shall be different than RDNA4, get_i shall be threadIdx.x % 16, get_j shall be  threadIdx.x % 16 + l.
- Just need two ggml_cuda_memcpy_1 for RDNA3 in load_generic, the position are get_i(0) * stride + get_j(0) and get_i(0) * stride + get_j(4)
- Use tile <16,16,half2> in mmf.cuh for RDNA3, make sure the right padding num as two memcpy might be executed at same time.
- Add tile<I,J,T> support for RDNA3 matrix C and everything shall be fine.

unverbraucht commented:
Closing this as long-since superseded
